### PR TITLE
Move settings to an app

### PR DESCRIPTION
**Pull request by ChristophWurst**
Todo

- [x] Namespaces
- [x] Autoloading (composer)
- [x] Translations
- [x] Release script
- [x] Webpack
- [x] Event names
- [x] Hooks
- [x] Test
- [x] Special settings magic
- [x] Background job registration
- [x] Activities
- [x] Find out what `help.php` does

### CODE REVIEW

**Comment by kesselb** [comment]
>  Find out what help.php does

What does help.php? :thinking:

**Comment by ChristophWurst** [comment]
> What does help.php? thinking

It helps :see_no_evil: 

Used here https://github.com/nextcloud/server/blob/master/settings/routes.php#L96, which still works.

**Comment by doliG** [comment]
I'm not sure to understand your git history 😅

**Comment by ChristophWurst** [comment]
> I'm not sure to understand your git history sweat_smile

https://dev.to/koffeinfrei/the-git-fixup-workflow-386d ;)

**Comment by doliG** [comment]
Wow, that's a nice trick, thanks for sharing.

Actually, what I don't get is why there is so little commits for so many changes (15 commits for 315 file changed).

It's very hard to understand what's going on here for someone who haven't followed the reface from the beginning 🤔

**Comment by ChristophWurst** [comment]
> Actually, what I don't get is why there is so little commits for so many changes (15 commits for 315 file changed).

I'm refactoring. And mostly moving stuff around. Eventually it will be only one commit.

> It's very hard to understand what's going on here for someone who haven't followed the reface from the beginning thinking

Sorry. We're just making the "settings" code an internal app. Which removes many hacks from the code and will allow us to clean up the server-side DI container registration to allow singletons of critical classes.

**Comment by ChristophWurst** [comment]
I think this is done. Works here and can't reproduce the random CI failures.

File: apps/provisioning_api/lib/Controller/UsersController.php [hunk 1]
```diff
@@ -61,7 +61,7 @@ class UsersController extends AUserData {
 	private $logger;
 	/** @var IFactory */
 	private $l10nFactory;
-	/** @var NewUserMailHelper */
+	/** @var \OCA\Settings\Mailer\NewUserMailHelper */
```
Comment (nickvergessen) [inline apps/provisioning_api/lib/Controller/UsersController.php]: Its imported? So should not be needed?

File: apps/settings/appinfo/info.xml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0"?>
+<info xmlns:xsi= "http://www.w3.org/2001/XMLSchema-instance"
+	  xsi:noNamespaceSchemaLocation="https://apps.nextcloud.com/schema/apps/info.xsd">
+	<id>settings</id>
+	<name>Nextcloud Settings</name>
+	<summary>Nextcloud Settings</summary>
+	<description>Nextcloud Settings</description>
+	<version>1.0.0</version>
+	<licence>agpl</licence>
+	<author>Bjoern Schiessle</author>
```
Comment (nickvergessen) [inline apps/settings/appinfo/info.xml]: *cough cough*

File: apps/settings/appinfo/info.xml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0"?>
+<info xmlns:xsi= "http://www.w3.org/2001/XMLSchema-instance"
+	  xsi:noNamespaceSchemaLocation="https://apps.nextcloud.com/schema/apps/info.xsd">
+	<id>settings</id>
+	<name>Nextcloud Settings</name>
```
Comment (nickvergessen) [inline apps/settings/appinfo/info.xml]:
```suggestion
	<name>Nextcloud settings</name>
```

File: apps/settings/appinfo/info.xml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+	<summary>Nextcloud Settings</summary>
+	<description>Nextcloud Settings</description>
+	<version>1.0.0</version>
+	<licence>agpl</licence>
+	<author>Bjoern Schiessle</author>
+	<namespace>Settings</namespace>
+	<default_enable/>
+
+	<category>files</category>
```
Comment (nickvergessen) [inline apps/settings/appinfo/info.xml]: :confused:

File: apps/settings/appinfo/info.xml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+	<bugs>https://github.com/nextcloud/server/issues</bugs>
+	<dependencies>
+		<nextcloud min-version="18" max-version="18"/>
+	</dependencies>
+</info>
```
Comment (nickvergessen) [inline apps/settings/appinfo/info.xml]: The settings sections and settings panels could be registered here now?

File: apps/settings/tests/Activity/SecurityProviderTest.php [hunk 6]
```diff
@@ -53,7 +53,7 @@ protected function setUp() {
 		$this->urlGenerator = $this->createMock(IURLGenerator::class);
 		$this->activityManager = $this->createMock(IManager::class);
 
-		$this->provider = new SecurityProvider($this->l10n, $this->urlGenerator, $this->activityManager);
+		$this->provider = new \OCA\Settings\Activity\SecurityProvider($this->l10n, $this->urlGenerator, $this->activityManager);
```
Comment (nickvergessen) [inline apps/settings/tests/Activity/SecurityProviderTest.php]: Same here, its imported, no need to use the full name

File: apps/settings/tests/Activity/CertificateControllerTest.php [hunk 7]
```diff
@@ -19,9 +19,9 @@
  *
  */
 
-namespace Tests\Settings\Controller;
+namespace OCA\Settings\Tests\Activity;
```
Comment (nickvergessen) [inline apps/settings/tests/Activity/CertificateControllerTest.php]: Should this be moved to `namespace OCA\Settings\Tests\Controller;`

Activity sounds like the wrong place 😅

File: apps/settings/appinfo/info.xml (hunk 5 above)
Comment (ChristophWurst) [reply apps/settings/appinfo/info.xml]: Yes, that can be done in a follow-up PR!

**Review by nickvergessen — APPROVED** [review]
Still works :elephant:

**Review by skjnldsv — CHANGES_REQUESTED** [review]
I cannot access the settings from the top right menu, I get redirected to /apps/files :)

**Comment by ChristophWurst** [comment]
> I cannot access the settings from the top right menu, I get redirected to /apps/files :)

is the app listed as enabled?

**Comment by skjnldsv** [comment]
> is the app listed as enabled?

Yes 
```
11:43 [admin@workstation] ~/Docker% docker-compose exec --user=docker php php occ app:list
Enabled:
  - accessibility: 1.4.0
  - activity: 2.11.0
  - admin_audit: 1.8.0
  - bookmarks: 2.1.1
  - calendar: 1.6.4
  - cloud_federation_api: 1.1.0
  - comments: 1.8.0
  - contacts: 3.1.4
  - dav: 1.14.0
  - federatedfilesharing: 1.8.0
  - federation: 1.8.0
  - files: 1.13.0
  - files_sharing: 1.10.0
  - files_texteditor: 2.10.0
  - files_trashbin: 1.8.0
  - files_versions: 1.11.0
  - files_videoplayer: 1.7.0
  - forms: 1.0.3
  - guests: 1.1.1
  - lookup_server_connector: 1.6.0
  - mail: 0.17.0
  - notifications: 2.6.0
  - oauth2: 1.6.0
  - pride: 0.0.3
  - privacy: 1.2.0
  - provisioning_api: 1.8.0
  - serverinfo: 1.8.0
  - settings: true
  - sharebymail: 1.8.0
  - spreed: 7.0.0-dev.2
  - systemtags: 1.8.0
  - testing: true
  - theming: 1.9.0
  - twofactor_backupcodes: 1.7.0
  - updatenotification: 1.8.0
  - viewer: 1.2.0
  - vueexample: 0.0.1
  - workflowengine: 2.0.0
Disabled:
  - NextcloudExtract
  - browser_warning
  - bruteforcesettings
  - circles
  - dashboard
  - encryption
  - extract
  - files_3d
  - files_accesscontrol
  - files_automatedtagging
  - files_external
  - files_retention
  - firstrunwizard
  - music
  - notes
  - password_policy
  - preferred_providers
  - previewgenerator
  - richdocuments
  - sandblast
  - tasks
  - terms_of_service
  - text
  - theming_customcss
  - twofactor_admin
  - twofactor_nextcloud_notification
  - twofactor_totp
  - user_ldap
```

I tried enabling again, and now the value is not `settings: true`, but `settings: 1.0.0`

**Comment by skjnldsv** [comment]
I get lots of errors:
```json
{
  "reqId": "0HBW0cdq3lHGqbG7EoX8",
  "level": 0,
  "time": "2019-09-26T09:49:13+00:00",
  "remoteAddr": "172.21.0.1",
  "user": "admin",
  "app": "PHP",
  "method": "GET",
  "url": "/core/preview?fileId=3614&c=bf943200b495ff4ca1bb00ed7b3d7dd2&x=250&y=250&forceIcon=0",
  "message": "Calling the \"Symfony\\Component\\EventDispatcher\\EventDispatcherInterface::dispatch()\" method with the event name as the first argument is deprecated since Symfony 4.3, pass it as the second argument and provide the event object as the first argument instead. at /var/www/html/3rdparty/symfony/event-dispatcher/EventDispatcher.php#58",
  "userAgent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.75 Safari/537.36",
  "version": "18.0.0.0"
}
```

**Comment by ChristophWurst** [comment]
> I get lots of errors:

Unrelated. Happens on master as well.

But I can confirm that the URLs do not work. Also settings vanished from the menu. @rullzer any idea? You fixed this once already :)

**Comment by rullzer** [comment]
Probably route caching.
Do you have memcache (apcu) configured?

**Comment by skjnldsv** [comment]
> Do you have memcache (apcu) configured?

yes, let me flush :toilet:

**Comment by rullzer** [comment]
Ok i have no idea what we did before to make it work
In any case. Pushed a fixup to fix it.

**Review by skjnldsv — APPROVED** [review]
Works! :)

**Comment by rullzer** [comment]
ci fails (checkers)

**Pull request merged**